iroh-net/src/hp/portmapper/pmp.rs

```diff
@@ -0,0 +1,301 @@
+#[repr(u8)]
+pub enum Opcode {
+    // 3.2.  Determining the External Address
+    DetermineExternalAddress = 0,
+    // 3.3.  Requesting a Mapping
+    MapUdp = 1,
+    // 3.3.  Requesting a Mapping
+    MapTcp = 2,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum MapProtocol {
+    UDP,
+    TCP,
+}
+
+#[derive(Debug)]
+pub enum Request {
+    ExternalAddress,
+    Mapping {
+        proto: MapProtocol,
+        local_port: u16,
+        external_port: u16,
+        lifetime_seconds: u32,
+    },
+}
+
+impl Request {
+    pub fn encode(&self) -> Vec<u8> {
+        match self {
+            Request::ExternalAddress => vec![
+                Version::NatPmp as u8,
+                Opcode::DetermineExternalAddress as u8,
+            ],
+            Request::Mapping {
+                proto,
+                local_port,
+                external_port,
+                lifetime_seconds,
+            } => {
+                let opcode = match proto {
+                    MapProtocol::UDP => Opcode::MapUdp,
+                    MapProtocol::TCP => Opcode::MapTcp,
+                };
+                let mut buf = vec![Version::NatPmp as u8, opcode as u8];
+                buf.extend_from_slice(&local_port.to_be_bytes());
+                buf.extend_from_slice(&external_port.to_be_bytes());
+                buf.extend_from_slice(&lifetime_seconds.to_be_bytes());
+                buf
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Response {
+    PublicAddress {
+        epoch_time: u32,
+        public_ip: std::net::Ipv4Addr,
+    },
+    PortMap {
+        proto: MapProtocol,
+        epoch_time: u32,
+        private_port: u16,
+        external_port: u16,
+        lifetime_seconds: u32,
+    },
+}
+
+/// Error ocurring when attempting to identity the [`Opcode`] in a server response.
+#[derive(Debug, PartialEq, Eq)]
+pub struct InvalidOpcode;
+
+impl TryFrom<u8> for Opcode {
+    type Error = InvalidOpcode;
+
+    fn try_from(value: u8) -> std::result::Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Opcode::DetermineExternalAddress),
+            1 => Ok(Opcode::MapUdp),
+            2 => Ok(Opcode::MapTcp),
+            _ => Err(InvalidOpcode),
+        }
+    }
+}
+
+/// Error ocurring when attempting to identify the [`Version`] in a server response.
+#[derive(Debug, PartialEq, Eq)]
+pub struct InvalidVersion;
+
+impl TryFrom<u8> for Version {
+    type Error = InvalidVersion;
+
+    fn try_from(value: u8) -> std::result::Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Version::NatPmp),
+            _ => Err(InvalidVersion),
+        }
+    }
+}
+
+/// Error ocurring when attempting to decode the [`ResultCode`] in a server response.
+#[derive(Debug, PartialEq, Eq)]
+pub struct InvalidResultCode;
+
+impl TryFrom<u16> for ResultCode {
+    type Error = InvalidResultCode;
+
+    fn try_from(value: u16) -> std::result::Result<Self, Self::Error> {
```

Comment:
This should make some of this code a little easier to write :) 
https://docs.rs/num_enum/